### PR TITLE
Friktion integration

### PR DESCRIPTION
**Pull request by sjillen**
- deposit    -> preview ok
- withdraw -> preview ok
- claim        -> preview ok

### CODE REVIEW

File: utils/validations.tsx [hunk 1]
```diff
@@ -188,6 +188,7 @@ export const validateBuffer = async (
   }
 };
 
+//TODO: delete this shit
```
Comment (Orelsanpls) [inline utils/validations.tsx]: :D

File: pages/dao/[symbol]/proposal/components/instructions/Friktion/Deposit.tsx [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    },
+  });
+
+  const [friktionVolts, setFriktionVolts] = useState<VoltList | null>(null);
+
+  const { ownedTokenAccountsInfo } = useGovernanceUnderlyingTokenAccounts(
+    governedAccountPubkey ?? undefined,
+  );
+
+  useEffect(() => {
```
Comment (Orelsanpls) [inline pages/dao/[symbol]/proposal/components/instructions/Friktion/Deposit.tsx]: maybe put this one in a `useFriktionVolt` file?

File: tools/sdk/friktion/friktion.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import { FriktionSnapshot, VoltSnapshot } from '@friktion-labs/friktion-sdk';
+
+const FRIKTION_SNAPSHOT_URL =
+  'https://friktion-labs.github.io/mainnet-tvl-snapshots/friktionSnapshot.json';
+
+// To get vaults that are not in circuit
+// Change to ?.filter((x) => !x.isInCircuits)
+
+// Now I have allowed only one vault -> 9cHT8d7d35ngj5i8WBZB8ibjnPLnvnym4tp4KoTCQtxw, our vault
```
Comment (Orelsanpls) [inline tools/sdk/friktion/friktion.ts]: Outdated comment

File: tools/sdk/friktion/friktion.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  const parsedResponse = (await response.json()) as FriktionSnapshot;
+  return parsedResponse.allMainnetVolts as VoltSnapshot[];
+};
+
+const createLabel = (voltType, tokenSymbol, vaultApy) =>
+  `Volt #${voltType} - ${VOLT_TYPES[voltType]} - ${tokenSymbol} - APY: ${vaultApy}%`;
+
+const formatVoltListFromSnapshot = (volts: VoltSnapshot[]): VoltList => {
+  return volts.reduce((list, snap) => {
+    if (!snap.voltVaultId || snap.voltType > 3) return list;
```
Comment (Orelsanpls) [inline tools/sdk/friktion/friktion.ts]: Why do we check that the `snap.voltType` is bigger than 3? To see if this is an unhandled type? Maybe use something based on the `VOLT_TYPES` variable.

```
!Object.keys(VOLT_TYPES).includes(snap.voltType)
```

File: tools/sdk/friktion/friktion.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  [vaultLabel: string]: VoltSnapshot;
+};
+
+const fetchVoltsSnapshot = async () => {
+  const response = await fetch(FRIKTION_SNAPSHOT_URL);
+  const parsedResponse = (await response.json()) as FriktionSnapshot;
+  return parsedResponse.allMainnetVolts as VoltSnapshot[];
+};
+
+const createLabel = (voltType, tokenSymbol, vaultApy) =>
```
Comment (Orelsanpls) [inline tools/sdk/friktion/friktion.ts]: typing? 🙏

Comment (sjillen) [reply tools/sdk/friktion/friktion.ts]: my bad ^^'

File: components/SelectOptionDetailed.tsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  text: string;
+  flag?: Flag;
+};
+
+type OptionDetails = {
+  [label: string]: OptionText;
+};
+
+const setFlagColorText = (flag?: Flag) => {
+  switch (flag) {
```
Comment (Orelsanpls) [inline components/SelectOptionDetailed.tsx]: You may create a map to set the colors. So typing would generate an error if the dev doesn't set a color when adding a new flag.

```
const flagColorClassMap: {
  [flag in Flag]: string;
} = {
  [Flag.Danger]: 'text-red',
  [Flag.Warning]: 'text-orange',
  [Flag.OK]: 'text-green',
};

const setFlagColorText = (flag?: Flag) => {
  if (!flag) {
    return '';
  }

  return flagColorClassMap[flag];
};
```

File: hooks/usefriktionVolts.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+      if (!governedAccountPubkey) return;
+
+      const volts = await getVolts({
+        connection,
+        wallet,
+        governancePubkey: governedAccountPubkey,
+      });
+      setFriktionVolts(volts);
+    })();
+  }, [governedAccountPubkey?.toBase58()]);
```
Comment (Orelsanpls) [inline hooks/usefriktionVolts.ts]: Should be enough I think:

```
}, [governedAccountPubkey]);
```

File: pages/dao/[symbol]/proposal/components/TokenAccountSelect.tsx [hunk 8]
```diff
@@ -1,16 +1,21 @@
 import { PublicKey } from '@solana/web3.js';
 import Select from '@components/inputs/Select';
 import { OwnedTokenAccountsInfo } from '@hooks/useGovernanceUnderlyingTokenAccounts';
+import SelectOptionDetailed, {
+  Flag,
+} from '../../../../../components/SelectOptionDetailed';
```
Comment (Orelsanpls) [inline pages/dao/[symbol]/proposal/components/TokenAccountSelect.tsx]: Should work:

```
import SelectOptionDetailed, { Flag } from '@components/SelectOptionDetailed';
```

File: pages/dao/[symbol]/proposal/components/TokenAccountSelect.tsx [hunk 9]
```diff
@@ -27,34 +32,23 @@ const TokenAccountSelect = ({
     const { mint, uiAmount, mintName, isATA } = ownedTokenAccountsInfo[
       pubkeyString
     ];
+    const details = {
```
Comment (Orelsanpls) [inline pages/dao/[symbol]/proposal/components/TokenAccountSelect.tsx]: Objects doesn't ensure the attributes order. Seems like to me that an array would be more appropriate:

```
const details = [
{ 
    label: 'Mint Name',
    text: mintName,
},
{ 
    label:  'UI Balance',
    text: uiAmount.toString(),
},
{ 
    label: 'Mint',
    text: mint.toBase58(),
},
];
```

File: components/SelectOptionDetailed.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+          <div
+            key={option.text}
+            className={`space-y-0.5 text-xs text-fgd-3 ${setFlagColorText(
+              option.flag,
+            )}`}
+          >
+            {`${label}: ${option.text}`}
+          </div>
+        ))}
+        {diffValue && (
```
Comment (Orelsanpls) [inline components/SelectOptionDetailed.tsx:53]: What the `diffValue` means please? I don't get it

File: pages/dao/[symbol]/proposal/components/instructions/Friktion/Deposit.tsx [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+import useFriktionVolt from '@hooks/usefriktionVolts';
+import useGovernanceUnderlyingTokenAccounts from '@hooks/useGovernanceUnderlyingTokenAccounts';
+import useInstructionFormBuilder from '@hooks/useInstructionFormBuilder';
+import { VoltData } from '@tools/sdk/friktion/friktion';
+import depositToVolt from '@tools/sdk/friktion/instructions/depositToVault';
+import { GovernedMultiTypeAccount } from '@utils/tokens';
+import { FriktionDepositForm } from '@utils/uiTypes/proposalCreationTypes';
+import SelectOptionDetailed, {
+  Flag,
+} from '../../../../../../../components/SelectOptionDetailed';
```
Comment (Orelsanpls) [inline pages/dao/[symbol]/proposal/components/instructions/Friktion/Deposit.tsx]:
```
import SelectOptionDetailed, { Flag } from '@components/SelectOptionDetailed';
```

File: pages/dao/[symbol]/proposal/components/instructions/Friktion/Deposit.tsx [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+import TokenAccountSelect from '../../TokenAccountSelect';
+
+const schema = yup.object().shape({
+  governedAccount: yup.object().required('Governance is required'),
+  sourceAccount: yup.string().typeError('Source account is required'),
+  volt: yup.string().required('Volt is required'),
+  uiAmount: yup.number().typeError('Amount is required'),
```
Comment (Orelsanpls) [inline pages/dao/[symbol]/proposal/components/instructions/Friktion/Deposit.tsx]: TypeError seems like a good thing to ensure its a number, good idea. Should you add `.required()` also?

File: pages/dao/[symbol]/proposal/components/instructions/Friktion/Withdraw.tsx [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+import useGovernanceUnderlyingTokenAccounts from '@hooks/useGovernanceUnderlyingTokenAccounts';
+import useInstructionFormBuilder from '@hooks/useInstructionFormBuilder';
+import { VoltData } from '@tools/sdk/friktion/friktion';
+import withdrawFromVault from '@tools/sdk/friktion/instructions/withdrawFromVault';
+import { uiAmountToNativeBN } from '@tools/sdk/units';
+import { GovernedMultiTypeAccount } from '@utils/tokens';
+import { FriktionWithdrawForm } from '@utils/uiTypes/proposalCreationTypes';
+import SelectOptionDetailed, {
+  Flag,
+} from '../../../../../../../components/SelectOptionDetailed';
```
Comment (Orelsanpls) [inline pages/dao/[symbol]/proposal/components/instructions/Friktion/Withdraw.tsx]:
```
import SelectOptionDetailed, { Flag } from '@components/SelectOptionDetailed';
```

File: pages/dao/[symbol]/proposal/components/instructions/Friktion/Withdraw.tsx [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+import TokenAccountSelect from '../../TokenAccountSelect';
+
+const schema = yup.object().shape({
+  governedAccount: yup.object().required('Governance is required'),
+  receiverAccount: yup.string().typeError('Source account is required'),
+  volt: yup.string().required('Volt is required'),
+  uiAmount: yup.number().typeError('Amount is required'),
```
Comment (Orelsanpls) [inline pages/dao/[symbol]/proposal/components/instructions/Friktion/Withdraw.tsx]: Same question about `required`

File: utils/uiTypes/proposalCreationTypes.ts [hunk 15]
```diff
@@ -36,17 +36,29 @@ export interface FormInstructionData {
 export interface SplTokenTransferForm {
   destinationAccount: string;
   amount: number | undefined;
-  governedTokenAccount: GovernedTokenAccount | undefined;
+  governedTokenAccount?: GovernedTokenAccount | undefined;
```
Comment (Orelsanpls) [inline utils/uiTypes/proposalCreationTypes.ts]: When it's optional, there is no need for `undefined`? You can go with:

```
export interface SplTokenTransferForm {
  destinationAccount?: string;
  amount?: number;
  governedTokenAccount?: GovernedTokenAccount;
  programId?: string;
  mintInfo?: MintInfo;
}
```

And in SplTokenTransfer:

```
  const [form, setForm] = useState<SplTokenTransferForm>({
    programId: programId?.toString(),
  });
```

File: utils/validations.tsx (hunk 1 above)
Comment (Orelsanpls) [reply utils/validations.tsx]: :D

**Review by Orelsanpls — CHANGES_REQUESTED** [review]
GG about Friktion!

File: utils/validations.tsx (hunk 1 above)
Comment (sjillen) [reply utils/validations.tsx]: :thumbup

File: components/SelectOptionDetailed.tsx (hunk 10 above)
Comment (sjillen) [reply components/SelectOptionDetailed.tsx:53]: bad variable naming

**Review by Orelsanpls — APPROVED** [review]
Let's go, let's try theses new instructions! 👍

**Comment by acamill** [comment]
Nice work sers

**Comment by acamill** [comment]
Damn there is a lot of code 😄

**Pull request merged**